src/main/resources/hudson/tasks/junit/History/index.jelly

```diff
@@ -24,84 +24,127 @@ THE SOFTWARE.
 
 <!-- Displays the chart that show how long builds are taking -->
 <?jelly escape-by-default='true'?>
-<j:jelly xmlns:j="jelly:core" xmlns:st="jelly:stapler" xmlns:l="/lib/layout" xmlns:bs="/bootstrap5" xmlns:c="/charts">
+<j:jelly xmlns:j="jelly:core" xmlns:st="jelly:stapler" xmlns:l="/lib/layout" xmlns:bs="/bootstrap5" xmlns:c="/charts" xmlns:f="/lib/form">
   <bs:layout title="${%title(it.testObject.displayName)}">
+    <j:set var="count" value="${it.asInt(request.getParameter('count'),100)}"/>
     <j:set var="start" value="${it.asInt(request.getParameter('start'),0)}"/>
-    <j:set var="end" value="${it.asInt(request.getParameter('end'),start+24)}"/>
+    <j:set var="end" value="${it.asInt(request.getParameter('end'),start+count-1)}"/>
+    <j:set var="historySummary" value="${it.retrieveHistorySummary(start, end)}"/>
     <st:adjunct includes="io.jenkins.plugins.echarts"/>
     <st:adjunct includes="io.jenkins.plugins.bootstrap5"/>
     <st:adjunct includes="io.jenkins.plugins.echarts-trend-default-setup"/>
+    <st:adjunct includes="hudson.tasks.junit.History.history"/>
     <link rel="stylesheet" href="${resURL}/plugin/junit/history/history.css"/>
 
     <st:once>
       <c:chart-setup id="test-history"/>
-      <script>var view =
-        <st:bind value="${it}"/>
-      </script>
-      <script type="text/javascript" src="${resURL}/plugin/junit/history/history.js"/>
     </st:once>
 
     <st:include from="${it.testObject}" it="${it.testObject}" page="sidepanel.jelly"/>
     <l:main-panel>
-      <j:choose>
-        <j:when test="${it.historyAvailable()}">
-          <bs:card title="${%History}" fontAwesomeIcon="chart-line">
-            <div id="trend-carousel" class="carousel slide carousel-dark" data-bs-interval="false">
-              <div class="carousel-inner">
-                <div class="carousel-item">
-                  <div id="test-duration-trend-chart" class="graph-cursor-pointer card-chart-carousel"/>
-                </div>
-                <div class="carousel-item active">
-                  <div id="test-result-trend-chart" class="graph-cursor-pointer card-chart-carousel"/>
-                </div>
-              </div>
-              <button class="carousel-control-prev" type="button" data-bs-target="#trend-carousel" data-bs-slide="prev">
-                <span class="carousel-control-prev-icon" aria-hidden="true"/>
-                <span class="visually-hidden">Previous</span>
-              </button>
-              <button class="carousel-control-next" type="button" data-bs-target="#trend-carousel" data-bs-slide="next">
-                <span class="carousel-control-next-icon" aria-hidden="true"/>
-                <span class="visually-hidden">Next</span>
-              </button>
-            </div>
-          </bs:card>
+      <div
+        id="history-data"
+        data-start="${start}"
+        data-end="${end}"
+        data-count="${count}"
+        data-trendChartJson="${historySummary.trendChartJson}"
```

Comment:
I'm not a big fan of this it bloats the page hugely and will slow down rendering I'd expect.

Ideally this would be loaded via JS, either by binding a JavaScriptMethod , similar to what was done before.

Or by using `fetch` to hit an API endpoint from JavaScript.

Reply:
I am not sure how to join the table and graphs into a single call for the fetch, any ideas? 

The JSON string can get into Megabytes in size; I am not a big fan of the data-* solution either. The inline JS felt more responsive on the UI side for big history window.

Reply:
Interestingly, it works way better if I do this

```
    <st:once>
      <c:chart-setup id="test-history"/>
      <script id="history-data-2" data-trendChartJson="${historySummary.trendChartJson}"></script>
    </st:once>
```

This seems to feel like before, but it is now a dummy empty script tag. Is this acceptable?

Reply:
Maybe it would more sense to invoke `${it.retrieveHistorySummary(start, end)}` directly on the JS side. (That would require the `bind` for `it` again.) You can get that data asynchronously as well to avoid a delay when the page is rendered.

Reply:
But how to render the table then? The point is that we should avoid doing the same computation twice and invoking it on the JS side won't provide the data back to Jelly view, or am I wrong?

Reply:
I see, I think your previous example would be ok but just set the type on the script tag so its compatible with CSP, see https://stackoverflow.com/a/45064686

Reply:
Wait a second, how does <st:bind ...> work? Will it recreate the HistoryTableResult every time I use it? Maybe I am not fully understanding how it works. I don't know if binding it keeps the instance alive via a session or something.

Reply:
The bind simply creates an Ajax endpoint proxy in the JS code so you can invoke code on the Java object that owns the view (i.e. `it`). It does not create objects, the object is already there, it is just not visible in JS. So the bind makes it visible (or callable).

Reply:
So, for every request on server side the History class is instantiated to process the request? Are the instances ever shared?

Reply:
I would expect not although you should be able to check in a debugger.

Generally for sharing data a static field is used or something like https://plugins.jenkins.io/caffeine-api/ which is likely overkill here but could be used.